[WGSL] ParameterizedTypeName should not parse the base

tadeuzagallo opened the pull request:
#### 4c7e11d0ab28f1c52e6cc6420e004d2306962480
<pre>
[WGSL] ParameterizedTypeName should not parse the base
<a href="https://bugs.webkit.org/show_bug.cgi?id=260611">https://bugs.webkit.org/show_bug.cgi?id=260611</a>
rdar://114321019

Reviewed by Dan Glastonbury.

Currently, when parsing, we make assumptions about what are the valid bases for
a ParameterizedTypeName, but that is not correct, as it can be affected by aliases
and also adds unnecessary complexity by going back and forth between identifiers
and enum values. To simplify things, we keep the base of a ParameterizedTypeName
as an identifier and also move type constructors from TypeStore into the regular
typing context. This makes it so that manipulating ParameterizedTypeNames works
more like any other type.

* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTTypeName.h:
(WGSL::AST::ParameterizedTypeName::base):
(WGSL::AST::ParameterizedTypeName::ParameterizedTypeName):
(WGSL::AST::ParameterizedTypeName::stringViewToKind): Deleted.
(WGSL::AST::ParameterizedTypeName::baseToString): Deleted.
(WGSL::AST::ParameterizedTypeName::base const): Deleted.
* Source/WebGPU/WGSL/ConstantRewriter.cpp:
(WGSL::ConstantRewriter::visit):
(WGSL::ConstantRewriter::materialize):
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::bindingMemberForGlobal):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseTypeNameAfterIdentifier):
(WGSL::Parser&lt;Lexer&gt;::parsePrimaryExpression):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::lookupType):
(WGSL::TypeChecker::vectorFieldAccess):
(WGSL::TypeChecker::allocateSimpleConstructor):
* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::TypeStore::TypeStore):
(WGSL::TypeStore::typeConstructorType):
(WGSL::TypeStore::constructType): Deleted.
(WGSL::TypeStore::allocateConstructor): Deleted.
* Source/WebGPU/WGSL/TypeStore.h:
* Source/WebGPU/WGSL/Types.cpp:
(WGSL::Type::dump const):
(WGSL::Type::size const):
(WGSL::Type::alignment const):
* Source/WebGPU/WGSL/Types.h:

Canonical link: <a href="https://commits.webkit.org/267348@main">https://commits.webkit.org/267348@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4081b13899a04644adfad581ecd2f6f094d92855

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16656 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18107 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15326 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16528 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19757 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16795 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17700 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16533 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16971 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14107 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18874 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14215 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14805 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21604 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15202 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14970 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19160 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15553 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13208 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14769 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3913 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19137 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15380 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
```diff
@@ -901,7 +901,8 @@ TEST(WGSLParserTests, RelationalExpression)
     testBinaryExpressionXY("x >= y"_s, WGSL::AST::BinaryOperation::GreaterEqual, { "x"_s, "y"_s });
-    testBinaryExpressionXY("x < y"_s,  WGSL::AST::BinaryOperation::LessThan,     { "x"_s, "y"_s });
+    // FIXME: implement template disambiguation
+    // testBinaryExpressionXY("x < y"_s, WGSL::AST::BinaryOperation::LessThan, { "x"_s, "y"_s });
```
This patch breaks "less than" relational expressions, but that's intentional. I'm writing a series of patches working towards having the correct template disambiguation, but the patches are fairly large, so I'm trying to break it into smaller bits.

djg commented on Source/WebGPU/WGSL/ConstantRewriter.cpp:
```diff
@@ -184,49 +184,16 @@ void ConstantRewriter::visit(AST::CallExpression& call)
-                return;
-            }
+        // FIXME: this implmentation of the vector constructors is incorrect, so
+        // the innefient comparison is irrelevant as it will be rewritten soon
```
typos: `implementation`, `inefficient`

djg reviewed (approved):
\o/

:shipit:

tadeuzagallo committed:
[WGSL] ParameterizedTypeName should not parse the base
https://bugs.webkit.org/show_bug.cgi?id=260611
rdar://114321019

Reviewed by Dan Glastonbury.

Currently, when parsing, we make assumptions about what are the valid bases for
a ParameterizedTypeName, but that is not correct, as it can be affected by aliases
and also adds unnecessary complexity by going back and forth between identifiers
and enum values. To simplify things, we keep the base of a ParameterizedTypeName
as an identifier and also move type constructors from TypeStore into the regular
typing context. This makes it so that manipulating ParameterizedTypeNames works
more like any other type.

* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTTypeName.h:
(WGSL::AST::ParameterizedTypeName::base):
(WGSL::AST::ParameterizedTypeName::ParameterizedTypeName):
(WGSL::AST::ParameterizedTypeName::stringViewToKind): Deleted.
(WGSL::AST::ParameterizedTypeName::baseToString): Deleted.
(WGSL::AST::ParameterizedTypeName::base const): Deleted.
* Source/WebGPU/WGSL/ConstantRewriter.cpp:
(WGSL::ConstantRewriter::visit):
(WGSL::ConstantRewriter::materialize):
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::bindingMemberForGlobal):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseTypeNameAfterIdentifier):
(WGSL::Parser<Lexer>::parsePrimaryExpression):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::TypeChecker):
(WGSL::TypeChecker::visit):
(WGSL::TypeChecker::lookupType):
(WGSL::TypeChecker::vectorFieldAccess):
(WGSL::TypeChecker::allocateSimpleConstructor):
* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::TypeStore::TypeStore):
(WGSL::TypeStore::typeConstructorType):
(WGSL::TypeStore::constructType): Deleted.
(WGSL::TypeStore::allocateConstructor): Deleted.
* Source/WebGPU/WGSL/TypeStore.h:
* Source/WebGPU/WGSL/Types.cpp:
(WGSL::Type::dump const):
(WGSL::Type::size const):
(WGSL::Type::alignment const):
* Source/WebGPU/WGSL/Types.h:

Canonical link: https://commits.webkit.org/267348@main
